### PR TITLE
TRC-28- Add `SearchBar` Component

### PR DESCRIPTION
**Pull request by mmcbride2929**
## TL/DR
This story involves creating a `SearchBar` component that includes an HTML input element and uses the CSS pseudo-class `::before` & `::after` to position the `searchIcon` & `closeIcon` inside it.

SEARCHBAR

## Overview of Changes

`SearchBar.tsx`
- Add Props interface with contents of `handleSearch`, `filterResults` & `customTestId` 
- Add `img` element for `searchIcon` 
- Add input `ref` that references input value
- Add and add `handleKeyDown` function to `input`
- Add `handleOnChange` function to `input`
- Replace input default close-button with `closeIcon` SVG
- Add `handleClick` function to `searchIcon` 

`SearchBar.scss`
- Style input element
- Use psuedo-class to style input for focus/hover
- Use psuedo-class `::before` to position the `searchIcon` 

`SearchBar.stories.tsx`
- Implements Story for `SearchBar` component using bind syntax

`SearchBar.test.tsx`
- Adds 5 tests: 
- SVG icon rendering + input element rendering, 
- onClick, onChange, & onKeyDown function call

`Icon.tsx`
- Tweak color hashes

### CODE REVIEW

File: apps/web/src/components/Icon/Icon.stories.tsx [hunk 1]
```diff
@@ -25,3 +26,9 @@ export const Home = () => (
     <HomeIcon />
   </Icon>
 );
+
+export const Search = (args: Props) => (
```
Comment (wijohnst) [inline apps/web/src/components/Icon/Icon.stories.tsx:30]: If you add the correct typing to this story you won't need to import the `Props` type as it is inferred from reference.

**Review by wijohnst — COMMENTED** [review]
Good start

File: apps/web/src/assets/homeIcon.svg
Comment (wijohnst) [inline apps/web/src/assets/homeIcon.svg:1]: I think I see why you made these changes to the SVG files. That said, they are a little out-of-scope from the original issue. 

When you find something to "fix" while you are implementing something else, you have a couple of options:

1. Fix them if they are small enough. But, make sure you talk about them in the PR description, since your reviewer might not have context for what you are actually fixing. 
2. Create a separate issue that is scoped to the changes.

I usually take the second route. 

Always remember that PRs that touch more files ALWAYS take longer to be reviewed. Keep your PRs as small as you can and as tightly scoped as you are comfortable with.

File: apps/web/src/components/Icon/Icon.stories.tsx [hunk 2]
```diff
@@ -1,7 +1,8 @@
-import Icon, { iconColor } from './Icon';
+import Icon, { iconColor, Props } from './Icon';
```
Comment (wijohnst) [inline apps/web/src/components/Icon/Icon.stories.tsx:1]: If you type this story correctly (`StoryFn`, `Meta`) you shouldn't need to bring in `Props` from `Icon`- the type should be inferred.

File: apps/web/src/components/SearchBar/SearchBar.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import * as React from 'react';
+import { useState } from 'react';
+import './SearchBar.scss';
+
+export interface Props {
+  /** What happens when we use a searchstring to search for results? */
+  handleSearch: (searchString: string) => void;
```
Comment (wijohnst) [inline apps/web/src/components/SearchBar/SearchBar.tsx:7]: Great use of the callback. I think you are starting to really understand separation of business logic from implementation logic.

File: apps/web/src/components/SearchBar/SearchBar.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  customTestId?: string;
+}
+
+// need to add jsdoc comment
+const SearchBar = ({
+  handleSearch,
+  customTestId = 'searchbar',
+}: Props): React.ReactElement => {
+  /** Tracks the string value within input */
+  const [searchString, setSearchString] = useState<string>('');
```
Comment (wijohnst) [inline apps/web/src/components/SearchBar/SearchBar.tsx:18]: Why are we storing `searchString` in local state? Think of `useState` kind of like `useEffect` - always have a good reason for using it, as it can make debugging or understanding the life cycle of a given component more difficult. 

You say that it "track the string value within input" - doesn't an `input` already do that?

File: apps/web/src/components/SearchBar/SearchBar.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  };
+
+  return (
+    <div className="searchbar-container" data-testid="searchbar-container">
+      <input
+        data-testid={customTestId}
+        type="search"
+        placeholder="Search Reddit"
+        value={searchString}
+        onChange={(e) => setSearchString(e.target.value)}
```
Comment (wijohnst) [inline apps/web/src/components/SearchBar/SearchBar.tsx:38]: Are you sure that we want to do this? What if I'm searching for "pie"?

I type "p" and the `onChange` event fires...
Then I type "i" and the `onChange` event fires...

Comment (wijohnst) [reply apps/web/src/components/SearchBar/SearchBar.tsx:38]: Went and looked at the actual input on reddit.com:

search_bar

So we do want to trigger a callback `onChange` but it isn't the same callback that we want to trigger `onKeyDown`.

File: apps/web/src/components/SearchBar/SearchBar.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+}: Props): React.ReactElement => {
+  /** Tracks the string value within input */
+  const [searchString, setSearchString] = useState<string>('');
+
+  /**
+   * Handles 'Enter' key press & invokes `handleSearch` func
+   *
+   * @param {React.KeyboardEvent} event - The keyboard event that triggered the function.
+   */
+  const handleKeyDown = (event: React.KeyboardEvent): void => {
```
Comment (wijohnst) [inline apps/web/src/components/SearchBar/SearchBar.tsx:25]: Is the `KeyboardEvent` the only event that can trigger search? 

What about a click event?

File: apps/web/src/components/SearchBar/SearchBar.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    if (event.key === 'Enter') {
+      handleSearch(searchString);
+    }
+  };
+
+  return (
+    <div className="searchbar-container" data-testid="searchbar-container">
```
Comment (wijohnst) [inline apps/web/src/components/SearchBar/SearchBar.tsx:32]: Is this wrapper `div` necessary?

File: apps/web/src/components/SearchBar/SearchBar.scss [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+input:hover,
+input:focus {
+  outline: 2px solid #3478cc;
+}
+
+input:focus {
+  border-radius: 30px 30px 0px 0px;
+}
```
[the remark this answers is not in the document]
Comment (wijohnst) [reply apps/web/src/components/SearchBar/SearchBar.scss:24]: Can any use of `px` be updated to a relative unit like `rem`?

File: apps/web/src/components/SearchBar/SearchBar.scss [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+.searchbar-container::after {
+  content: url('../../assets/searchIcon.svg');
```
[the remark this answers is not in the document]
Comment (wijohnst) [reply apps/web/src/components/SearchBar/SearchBar.scss:27]: Just something to think about:

What if we had the requirement that the `searchIcon` was a different color based on `:focus` state? Like it should be rendered in a lighter color if it's not focused?

**Review by wijohnst — COMMENTED** [review]
Great start!

File: apps/web/src/components/SearchBar/SearchBar.tsx (hunk 7 above)
Comment (mmcbride2929) [reply apps/web/src/components/SearchBar/SearchBar.tsx:32]: In order to get the icon to show up, I needed to wrap the input with a div, and use `.searchbar-container::before` to target & add the SVG. If I used `input-container::before` the icon wouldn't appear.

File: apps/web/src/components/SearchBar/SearchBar.tsx (hunk 6 above)
Comment (mmcbride2929) [reply apps/web/src/components/SearchBar/SearchBar.tsx:25]: As for the `searchIcon` & `closeIcon` I have added and positioned them both, using ::before & ::after. After doing so, I was looking up how to add an onClick to clear the input & trigger handleSearch, but I haven't been able to find much besides this workaround on stack-overflow. Wanted to see what you thought.

#TLDR
"This is not possible; pseudo-elements are not part of the DOM at all so you can't bind any events directly to them, you can only bind to their parent elements.

If you must have a click handler on the red region only, you have to make a child element, like a span, place it right after the opening <p> tag, apply styles to p span instead of p:before, and bind to it."

https://stackoverflow.com/questions/7478336/only-detect-click-event-on-pseudo-element

File: apps/web/src/components/SearchBar/SearchBar.tsx (hunk 7 above)
Comment (mmcbride2929) [reply apps/web/src/components/SearchBar/SearchBar.tsx:87]: I added the div because without it I could not get the `searchIcon` or `closeIcon` to appear. I also had a tough time figuring out how to give these icons onClick events but eventually found this solution/workaround.

https://stackoverflow.com/questions/7478336/only-detect-click-event-on-pseudo-element

"This is not possible; pseudo-elements are not part of the DOM at all so you can't bind any events directly to them, you can only bind to their parent elements.

If you must have a click handler on the red region only, you have to make a child element, like a span, place it right after the opening <p> tag, apply styles to p span instead of p:before, and bind to it."

File: apps/web/src/components/SearchBar/SearchBar.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (mmcbride2929) [reply apps/web/src/components/SearchBar/SearchBar.scss:34]: I could not figure this out. I tried adding an opacity of 0.75 to the icons, and :hover an opacity of 1, but this did not work.

File: apps/web/src/components/SearchBar/SearchBar.tsx (hunk 6 above)
Comment (mmcbride2929) [reply apps/web/src/components/SearchBar/SearchBar.tsx]: I need to figure out how to get the `closeButton` to be hidden when the input is empty, and how to hide it when the `closeButton` is pressed & the input is cleared. I tried conditionally rendering the `span` when the ref is `null` equal to an empty string but it did not work.

File: apps/web/src/components/SearchBar/SearchBar.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (wijohnst) [reply apps/web/src/components/SearchBar/SearchBar.scss:34]: Good things to try for sure. Same thing that I did when I had to do this for work. 

I'll put the detailed solution into Slack. Don't let me forget.

File: apps/web/src/components/SearchBar/SearchBar.tsx (hunk 7 above)
Comment (wijohnst) [reply apps/web/src/components/SearchBar/SearchBar.tsx:87]: The "pseudo-element are not part of the DOM bit" is something that I always forget. (Coincidently, this is very similar to the reason you can't dynamically style the content of `:before` and `:after` pseudo-elements like styling most other things in the DOM.)

Thank you for the explanation. Consider my comment resolved.

File: apps/web/src/components/SearchBar/SearchBar.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (willcary) [reply apps/web/src/components/SearchBar/SearchBar.scss:34]: I noticed the input loses its outline (hover state) when you hover over the search icon. It seems like the hover outline should be maintained. I'm not convinced this is the best solution, but you could use the `.searchbar-container` div for the outline/border instead of the input.

[the remark this answers is not in the document]
Comment (willcary) [reply apps/web/src/components/SearchBar/SearchBar.scss:34]: I think if you're using html elements (`span`) for the search and close icons, you don't need to use the pseudo elements (::before and ::after). As I understand it, their whole function is to add styling/content to a page without adding html to the DOM. I see in another resolved conversation that the pseudo elements weren't interactive because they aren't actually in the DOM which is why you used spans so totally understandable, but it might make sense to refactor to a different approach now that we know those pseudo-elements don't suit this situation. Maybe Will J. can chime in if I'm off base here.

**Review by willcary — COMMENTED** [review]
Nice work, Michael! This is a tricky component to craft. I left a few comments to look over and see if they're worth incorporating. If you'd like to do some pair programming to try to solve some issues (like the hiding the close button when the input is empty), I'm down to try and help.

File: apps/web/src/components/SearchBar/SearchBar.test.tsx
Comment (wijohnst) [inline apps/web/src/components/SearchBar/SearchBar.test.tsx:1]: Just thinking out loud, but what happens if we click or keydown and don't have a "searchString"?

File: apps/web/src/components/SearchBar/SearchBar.tsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  /** What happens when a user begins typing what they're searching for? */
+  filterResults: (searchString: string) => object[];
+  /** An optional custom test ID for the component.  */
+  customTestId?: string;
+}
+
+/**
+ * A component that renders a SearchBar
+ *
+ * @param {Object} props - The component props
```
Comment (wijohnst) [inline apps/web/src/components/SearchBar/SearchBar.tsx:18]: Just a note, but you can type the JSDocs more tightly. 
```suggestion
 * @param {Props} props - The component props
```

File: apps/web/src/components/SearchBar/SearchBar.test.tsx
Comment (wijohnst) [reply apps/web/src/components/SearchBar/SearchBar.test.tsx:1]: Never mind. I see that you already handled this. 

Not necessary, but we could add a test case for this.

**Review by wijohnst — APPROVED** [review]
Really really strong work on this. This is professional level code. I'm very impressed with your rapid progress! 🚀

File: apps/web/src/components/SearchBar/SearchBar.tsx [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+   */
+  const handleKeyDown = (
+    event: React.KeyboardEvent<HTMLInputElement>
+  ): void => {
+    if (event.key === 'Enter') {
+      const searchString = ref.current?.value;
+      if (
+        searchString !== undefined &&
+        searchString !== null &&
+        searchString !== ''
```
Comment (willcary) [inline apps/web/src/components/SearchBar/SearchBar.tsx:46]: Since you use these sets of conditions 3 times would it make sense to save them in a variable for reuse? I think it's fine and maybe clearer the way as it is though.

**Review by willcary — APPROVED** [review]
Awesome job Michael! I learned quite a bit reviewing your tests. Lots of new methods I didn't know before. I left one comment, but am totally fine if you disregard it.

File: apps/web/src/components/SearchBar/SearchBar.tsx (hunk 11 above)
Comment (wijohnst) [reply apps/web/src/components/SearchBar/SearchBar.tsx:46]: This is like the only time I recommend using the `!!` operator: 

Screen Shot 2023-04-02 at 3 12 13 PM

Comment (wijohnst) [reply apps/web/src/components/SearchBar/SearchBar.tsx:46]: So this  could all be simplified to `!!searchString`. If you want something more semantic you could return a boolean from a function like:

```ts
const isSearchStringValud = (searchString : string) : boolean => return !!searchString;
```
